exchange/bitswap/bitswap.go

```diff
@@ -270,26 +270,40 @@ func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg
 	// TODO: this is bad, and could be easily abused.
 	// Should only track *useful* messages in ledger
 
-	if len(incoming.Blocks()) == 0 {
+	iblocks := incoming.Blocks()
+
+	if len(iblocks) == 0 {
 		return
 	}
 
 	// quickly send out cancels, reduces chances of duplicate block receives
 	var keys []u.Key
-	for _, block := range incoming.Blocks() {
+	for _, block := range iblocks {
 		keys = append(keys, block.Key())
 	}
 	bs.wm.CancelWants(keys)
 
-	for _, block := range incoming.Blocks() {
+	for _, block := range iblocks {
 		bs.counterLk.Lock()
 		bs.blocksRecvd++
-		if has, err := bs.blockstore.Has(block.Key()); err == nil && has {
+		has, err := bs.blockstore.Has(block.Key())
```

Comment:
> worst case its a disk lookup

that's pretty expensive to do then-- this function is called per message received. it may block other things, and it may make sense to do all disk accesses together (i.e. send a bunch of read calls to the OS) instead of doing it RTT style.

maybe this should have the same sort of thing that "add" has-- send on a channel to a long queue that then goes through and calls `.Has` sequentially. not sure of all the implications-- `Has` is how it gets to the client program who may have requested this block, too, right? we want that to happen asap? (but the other `Has` things-- like maybe sending it out to peers-- can wait.)